Fix Key Verification and its test

TobiasFella opened the pull request:
- Add outgoing sessions to the list as well
- Verify that session wasn't cancelled
- Test session between different accounts in order to not mess up the database
- Make sure we have the key for the other party by being in an encrypted room
- Make it possible to verify keys of other users
- Capture QPointer by reference to correctly share the session

TobiasFella committed:
Fix Key Verification and its test

- Add outgoing sessions to the list as well
- Verify that session wasn't cancelled
- Test session between different accounts in order to not mess up the database
- Make sure we have the key for the other party by being in an encrypted room
- Make it possible to verify keys of other users
- Capture QPointer by reference to correctly share the session

KitsuneRal commented on autotests/testkeyverification.cpp:
```diff
@@ -6,6 +6,8 @@
 #include <QTest>
 #include "testutils.h"
 #include <qt_connection_util.h>
+#include <QTimer>
```
```suggestion
#include <QtCore/QTimer>
```

KitsuneRal commented on autotests/testkeyverification.cpp:
```diff
@@ -15,23 +17,37 @@ private Q_SLOTS:
+        CREATE_CONNECTION(b, "alice2", "secret", "AlicePhone")
+
+        a->requestDirectChat(b->userId());
+            connectSingleShot(b.get(), &Connection::invitedRoom, this, [=](Quotient::Room* room, Quotient::Room*) {
```
```suggestion
        connectSingleShot(b.get(), &Connection::invitedRoom, this, [=](Quotient::Room* room) {
```
You can omit trailing signal parameters you don't use in the slot.

KitsuneRal commented on autotests/testkeyverification.cpp:
```diff
@@ -15,23 +17,37 @@ private Q_SLOTS:
         connect(a.get(), &Connection::newKeyVerificationSession, this, [&](KeyVerificationSession* session) {
             aSession = session;
+            QVERIFY(aSession);
+            connect(session, &KeyVerificationSession::stateChanged, this, [session](){
```
```suggestion
            connect(session, &KeyVerificationSession::stateChanged, this, [session] {
```

KitsuneRal commented on autotests/testkeyverification.cpp:
```diff
@@ -15,23 +17,37 @@ private Q_SLOTS:
         });
-        a->startKeyVerificationSession(b->deviceId());
-        connect(b.get(), &Connection::newKeyVerificationSession, this, [=](KeyVerificationSession* session) {
+        QTimer::singleShot(3000, this, [a, b](){
```
Can I have a comment in the code here on why the delay is needed?

Also:
```suggestion
        QTimer::singleShot(3000, this, [a, b] {
```

KitsuneRal commented on autotests/testkeyverification.cpp:
```diff
@@ -15,23 +17,37 @@ private Q_SLOTS:
+            a->startKeyVerificationSession(b->userId(), b->deviceId());
+        });
+        connect(b.get(), &Connection::newKeyVerificationSession, this, [&](KeyVerificationSession* session) {
+            connect(session, &KeyVerificationSession::stateChanged, this, [session](){
```
```suggestion
            connect(session, &KeyVerificationSession::stateChanged, this, [session] {
```

KitsuneRal commented on autotests/testkeyverification.cpp:
```diff
@@ -44,15 +60,18 @@ private Q_SLOTS:
                     QVERIFY(aSession);
                     QVERIFY(aSession->sasEmojis() == session->sasEmojis());
                     session->sendMac();
+                    aSession->sendMac();
```
Soooo, does it mean that success of previous test runs was due to the test missing some important validation? :) In other words - should this test also check that upon `&KeyVerificationSession::finished` the respective party has received the MAC?

TobiasFella replied:
i don't think this change is particularly important, it was something i did while trying to figure out why the QPointer is nullptr before finding the actual reason

TobiasFella replied:
yes, checking for the mac actually being received makes sense, i'll add that

KitsuneRal reviewed (changes requested):
I have questions :)

TobiasFella committed:
Update autotests/testkeyverification.cpp

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Update autotests/testkeyverification.cpp

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Update autotests/testkeyverification.cpp

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Update autotests/testkeyverification.cpp

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Update autotests/testkeyverification.cpp

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Revert irrelevant change and add comment

TobiasFella committed:
Replace tests with more atomic unit tests and allow for more state transitions in the KeyVerificationSession

The old test was largely unmaintainable and is thus being replaced by actual unit tests

Since we currently don't have a way of getting the events sent by a connection in a test, we are somewhat limited in what we can test.
Add several TODOs for when this limitation is removed

TobiasFella commented:
Updated with more stable - but also more limited - tests

KitsuneRal commented on autotests/testkeyverification.cpp:
```diff
@@ -5,54 +5,61 @@
 #include <QTest>
 #include "testutils.h"
-#include <qt_connection_util.h>
+#include <QDateTime>
```
```suggestion
#include <QtCore/QDateTime>
```

KitsuneRal commented on lib/connection.cpp:
```diff
@@ -2485,3 +2486,8 @@ bool Connection::isVerifiedSession(const QString& megolmSessionId) const
+void Connection::completeSetup(const QString &mxId)
+{
+    d->completeSetup(mxId);
+}
```
Given that you always create a `Connection` object just before calling this wrapper, it makes sense to have a static factory instead:
```suggestion
Connection* Connection::makeMockConnection(const QString& mxId)
{
    auto* c = new Connection;
    c->d->completeSetup(mxId);
    return c;
}
```

KitsuneRal commented on lib/connection.h:
```diff
@@ -700,11 +700,14 @@ public Q_SLOTS:
 #endif
 
+    // WARNING: Use only for unit tests! This will not lead to a correctly working connection!
+    void completeSetup(const QString& mxId);
```
...and then you don't need that warning :)
```suggestion
     static Connection* makeMockConnection(const QString& mxId);
```

KitsuneRal reviewed (changes requested):
Quite a clever shortcut with `Connection`, I like it! Suggest just one simple improvement but the approach looks unexpectedly reasonable.

TobiasFella committed:
Apply suggestions from code review

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Fix
